src/java.base/share/classes/java/util/concurrent/CopyOnWriteArrayList.java

```diff
@@ -263,7 +265,7 @@ public int indexOf(@Nullable Object o) {
      *         {@code -1} if the element is not found.
      * @throws IndexOutOfBoundsException if the specified index is negative
      */
-    public int indexOf(E e, int index) {
+    public int indexOf(@Nullable E e, int index) {
```

Comment:
Why is this change desirable (same for `lastIndexOf` below)? If the collection only accepts non-null elements, why allow checking with a null value? It won't throw an NPE, but it will always be false.
The `indexOf` variant that takes an `Object` needs to take an `@Nullable Object` parameter to allow both instantiations of the collection, but that doesn't apply here.

If you think they are indeed correct, having a [`@CFComment`](https://github.com/eisop/checker-framework/blob/master/checker-qual/src/main/java/org/checkerframework/framework/qual/CFComment.java) would be useful.

Reply:
Part of my goal here is consistency, too: `contains` already has a `@Nullable` parameter. So let's make them all `@Nullable` or all _not_ `@Nullable`.

The broader question is still a good one. As in [our `retainAll` discussion](https://github.com/eisop/jdk/pull/9#discussion_r862885363), I've been operating under the principle of "accept the most general type that can't lead to NPE," but of course we're well aware that the principles here are controversial :) (That's not necessarily even my personal favored principle.)

I probably should have sent the `poll` and `peek` changes (which probably actually affect users) separately from the more complicated changes. It would be nice to get those in. But for the changes we've been discussing, I don't think we're affected much by what we decide to do, so we can continue discussing or just drop them, as you prefer.

Reply:
`contains` takes an `Object` parameter, so that discussion should go along with the other `Object` vs. `@Nullable Object` methods.
For methods that take the type parameter, I don't see a benefit in allowing null. I agree that it wouldn't cause an NPE, but it would just allow calls that will return false. Is there a particular pattern where this would be much more convenient?

Reply:
Ah, thanks, I had forgotten about the distinction of `Object` vs. `E`. I am a bit surprised that they use `E` here, given the precedent of `List.indexOf`, but of course there are good arguments for not permitting calls that will always return `false` (compare https://errorprone.info/bugpattern/CollectionIncompatibleType). And if their goal is to use `E` to forbid such calls, then it's natural to extend that argument to keeping the type as `E` instead of `@Nullable E`.

I _might_ still lean toward `@Nullable E` if we were writing The Official JSpecify Stubs: The theory would be that, if we didn't, tools could choose to insert runtime null checks for code that calls `indexOf` on a `CopyOnWriteArrayList<@NonNull String>`. But my recollection is that the only tool in common use that inserts such checks, Kotlin, does _not_ insert checks in that particular case.

But for Checker Framework users (and for all non-hypothetical use cases), I'm not aware of examples in which `@Nullable` would improve anyone's life.